rocAL/rocAL/source/pipeline/master_graph.cpp

```diff
@@ -704,51 +814,139 @@ void MasterGraph::box_encoder(std::vector<float> &anchors, float criteria, const
+std::vector<rocalTensorList *> MasterGraph::create_caffe2_lmdb_record_meta_data_reader(const char *source_path, MetaDataReaderType reader_type , MetaDataType label_type)
 {
     if( _meta_data_reader)
         THROW("A metadata reader has already been created")
+    if (_augmented_meta_data)
+        THROW("Metadata output already defined, there can only be a single output for metadata augmentation")
+
     MetaDataConfig config(label_type, reader_type, source_path);
     _meta_data_graph = create_meta_data_graph(config);
-    _meta_data_reader = create_meta_data_reader(config);
-    _meta_data_reader->init(config);
+    _meta_data_reader = create_meta_data_reader(config, _augmented_meta_data);
     _meta_data_reader->read_all(source_path);
-    if (_augmented_meta_data)
-        THROW("Metadata output already defined, there can only be a single output for metadata augmentation")
-    else
-        _augmented_meta_data = _meta_data_reader->get_output();
-    return _meta_data_reader->get_output();
+    if(reader_type == MetaDataReaderType::CAFFE2_META_DATA_READER)
+    {
+        std::vector<size_t> dims = {1};
+        auto default_labels_info  = rocalTensorInfo(std::move(dims), _mem_type, RocalTensorDataType::INT32); // Create default labels Info
+        default_labels_info.set_metadata();
+        _meta_data_buffer_size.emplace_back(_user_batch_size * sizeof(vx_int32));
+
+        for(unsigned i = 0; i < _user_batch_size; i++)
+        {
+            auto info = default_labels_info;
+            auto tensor = new rocalTensor(info);
+            _labels_tensor_list.push_back(tensor);
+        }
+        _metadata_output_tensor_list.emplace_back(&_labels_tensor_list);
+    }
+    else if(reader_type == MetaDataReaderType::CAFFE2_DETECTION_META_DATA_READER)
+    {
+        std::vector<size_t> dims = { MAX_OBJECTS };
+        auto default_labels_info  = rocalTensorInfo(std::move(dims), _mem_type, RocalTensorDataType::INT32); // Create default labels Info
+        default_labels_info.set_metadata();
+        _meta_data_buffer_size.emplace_back(_user_batch_size * default_labels_info.data_size());
+
+        dims = { MAX_OBJECTS, BBOX_COUNT };
+        auto default_bbox_info  = rocalTensorInfo(std::move(dims), _mem_type, RocalTensorDataType::FP32); // Create default Bbox Info
+        default_bbox_info.set_metadata();
+        _meta_data_buffer_size.emplace_back(_user_batch_size * default_bbox_info.data_size());
+
+        for(unsigned i = 0; i < _user_batch_size; i++)
+        {
+            auto labels_info = default_labels_info;
+            auto bbox_info = default_bbox_info;
+            _labels_tensor_list.push_back(new rocalTensor(labels_info));
+            _bbox_tensor_list.push_back(new rocalTensor(bbox_info));
+        }
+        _metadata_output_tensor_list.emplace_back(&_labels_tensor_list);
```

Comment:
please return _labels_tensor_list and _bbox_tensor_list directly. Applies to all such code in this file

Reply:
Since we wanted to return a single data struct we are adding the reference to the different metadata TensorList and passing to the user